File: src/main/java/page/support/AbJTable.java
package page.support;

import javax.annotation.Nonnull;
import javax.swing.*;
import javax.swing.event.TableModelListener;
import javax.swing.table.DefaultTableColumnModel;
import javax.swing.table.TableColumn;
import javax.swing.table.TableModel;

public abstract class AbJTable extends JTable implements TableModel {

	private static final long serialVersionUID = 1L;

	protected final int[] lnk;
	protected final String[] title;

	private final TModel tm = new TModel(this);

	protected AbJTable(@Nonnull String[] title) {
		this.title = title;

		lnk = new int[getColumnCount()];

		for (int i = 0; i < lnk.length; i++)
			lnk[i] = i;

		setColumnModel(tm);
		setModel(this);
		setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
	}

	@Override
	public void addTableModelListener(TableModelListener arg0) {
	}

	@Override
	public abstract Class<?> getColumnClass(int c);

	@Override
	public int getColumnCount() {
		return title.length;
	}

	@Override
	public abstract String getColumnName(int c);

	@Override
	public abstract int getRowCount();

	@Override
	public abstract Object getValueAt(int r, int c);

	@Override
	public boolean isCellEditable(int arg0, int arg1) {
		return false;
	}

	@Override
	public void removeTableModelListener(TableModelListener arg0) {
	}

	@Override
	public void setValueAt(Object arg0, int arg1, int arg2) {
	}

	protected void swap(int c, int nc) {
		int t = lnk[nc];
		lnk[nc] = lnk[c];
		lnk[c] = t;
	}

}

class TModel extends DefaultTableColumnModel {

	private static final long serialVersionUID = 1L;

	private final AbJTable t;

	protected TModel(AbJTable table) {
		t = table;
	}

	@Override
	public void moveColumn(int c, int nc) {
		c = Math.max(c, 0);
		c = Math.min(c, getColumnCount() - 1);
		nc = Math.max(nc, 0);
		nc = Math.min(nc, getColumnCount() - 1);

		if (c == nc)
			return;
		t.swap(c, nc);
		super.moveColumn(c, nc);

	}

	@Override
	public void removeColumn(TableColumn tc) {
	}

}

File: src/page/support/AbJTable.kt
package page.support

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.UnitLevel
import io.BCPlayer
import page.JL
import page.anim.AnimBox
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import javax.swing.JTable
import javax.swing.ListSelectionModel
import javax.swing.event.TableModelListener
import javax.swing.table.DefaultTableColumnModel
import javax.swing.table.TableColumn
import javax.swing.table.TableModel

abstract class AbJTable protected constructor() : JTable(), TableModel {
    protected val lnk: IntArray
    private val tm = TModel(this)
    override fun addTableModelListener(arg0: TableModelListener) {}
    abstract fun getColumnClass(c: Int): Class<*>?
    abstract fun getColumnCount(): Int
    abstract fun getColumnName(c: Int): String
    abstract fun getRowCount(): Int
    abstract fun getValueAt(r: Int, c: Int): Any
    override fun isCellEditable(arg0: Int, arg1: Int): Boolean {
        return false
    }

    override fun removeTableModelListener(arg0: TableModelListener) {}
    override fun setValueAt(arg0: Any, arg1: Int, arg2: Int) {}
    fun swap(c: Int, nc: Int) {
        val t = lnk[nc]
        lnk[nc] = lnk[c]
        lnk[c] = t
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        lnk = IntArray(getColumnCount())
        for (i in lnk.indices) lnk[i] = i
        setColumnModel(tm)
        setModel(this)
        setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
    }
}

internal class TModel(private val t: AbJTable) : DefaultTableColumnModel() {
    override fun moveColumn(c: Int, nc: Int) {
        if (c == nc) return
        t.swap(c, nc)
        super.moveColumn(c, nc)
    }

    override fun removeColumn(tc: TableColumn) {}

    companion object {
        private const val serialVersionUID = 1L
    }
}
